server/repository/src/db_diesel/assets/asset.rs

```diff
@@ -0,0 +1,271 @@
+                }
+                AssetSortField::ReplacementDate => {
+                    apply_sort!(query, sort, asset_dsl::replacement_date)
+                }
+                AssetSortField::ModifiedDatetime => {
+                    apply_sort!(query, sort, asset_dsl::modified_datetime)
+                }
+            }
+        } else {
+            query = query.order(asset_dsl::id.asc())
+        }
+
+        let final_query = query
+            .offset(pagination.offset as i64)
+            .limit(pagination.limit as i64);
+
+        // Debug diesel query
+        // println!(
+        //    "{}",
+        //     diesel::debug_query::<DBType, _>(&final_query).to_string()
+        // );
+
+        let result = final_query.load::<Asset>(&self.connection.connection)?;
+
+        Ok(result.into_iter().map(to_domain).collect())
+    }
+}
+
+fn to_domain(asset_row: AssetRow) -> Asset {
+    asset_row
+}
+
+type BoxedAssetQuery = IntoBoxed<'static, asset::table, DBType>;
+
+fn create_filtered_query(filter: Option<AssetFilter>) -> BoxedAssetQuery {
+    let mut query = asset_dsl::asset.into_boxed();
+
+    if let Some(f) = filter {
+        let AssetFilter {
+            id,
+            serial_number,
+            category_id,
+            type_id,
+            catalogue_item_id,
+            installation_date,
+            replacement_date,
+        } = f;
+
+        apply_equal_filter!(query, id, asset_dsl::id);
+        apply_string_filter!(query, serial_number, asset_dsl::serial_number);
+        apply_equal_filter!(query, category_id, asset_dsl::asset_category_id);
+        apply_equal_filter!(query, type_id, asset_dsl::asset_type_id);
+        apply_equal_filter!(query, catalogue_item_id, asset_dsl::asset_catalogue_item_id);
+        apply_date_filter!(query, installation_date, asset_dsl::installation_date);
+        apply_date_filter!(query, replacement_date, asset_dsl::replacement_date);
+    }
+    query.filter(asset_dsl::deleted_datetime.is_null()) // Don't include any deleted items
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        assets::{
+            asset::AssetRepository,
+            asset_category_row::{AssetCategoryRow, AssetCategoryRowRepository},
+            asset_class_row::{AssetClassRow, AssetClassRowRepository},
+            asset_row::{AssetRow, AssetRowRepository},
+            asset_type_row::{AssetTypeRow, AssetTypeRowRepository},
+        },
+        mock::{mock_store_a, MockDataInserts},
+        test_db, EqualFilter,
+    };
+
+    use super::AssetFilter;
+
+    #[actix_rt::test]
+    async fn test_asset_query_repository() {
+        // Prepare
+        let (_, storage_connection, _, _) = test_db::setup_all(
+            "test_asset_query_repository",
+            MockDataInserts::none().stores(),
+        )
+        .await;
+
+        // TODO: Replace this reference data with mock data, or inserted data from https://github.com/msupply-foundation/open-msupply/issues/3035
```

Comment:
Haven't done this as since we'd doing migrations, it seems a good idea to wait...